includes/wc-payment-api/class-wc-payments-api-client.php

```diff
@@ -2088,20 +2089,34 @@ protected function request( $params, $api, $method, $is_site_specific = true, $u
 			);
 		}
 
-		$response = $this->http_client->remote_request(
-			[
-				'url'             => $url,
-				'method'          => $method,
-				'headers'         => apply_filters( 'wcpay_api_request_headers', $headers ),
-				'timeout'         => self::API_TIMEOUT_SECONDS,
-				'connect_timeout' => self::API_TIMEOUT_SECONDS,
-			],
-			$body,
-			$is_site_specific,
-			$use_user_token
-		);
+		$headers        = apply_filters( 'wcpay_api_request_headers', $headers );
+		$stop_trying_at = time() + self::API_TIMEOUT_SECONDS;
+		$retries        = 0;
+		$retries_limit  = array_key_exists( 'Idempotency-Key', $headers ) ? self::API_RETRIES_LIMIT : 0;
 
-		$response = apply_filters( 'wcpay_api_request_response', $response, $method, $url, $api );
+		while ( true ) {
+			$response = $this->http_client->remote_request(
+				[
+					'url'             => $url,
+					'method'          => $method,
+					'headers'         => $headers,
+					'timeout'         => self::API_TIMEOUT_SECONDS,
+					'connect_timeout' => self::API_TIMEOUT_SECONDS,
+				],
+				$body,
+				$is_site_specific,
+				$use_user_token
+			);
+
+			$response      = apply_filters( 'wcpay_api_request_response', $response, $method, $url, $api );
+			$response_code = wp_remote_retrieve_response_code( $response );
```

Comment:
It seems that relying on this is not accurate here. Looking futhrer downstream of `WC_Payments_Http::remote_request`, I notice that we throw `Connection_Exception` in  `WC_Payments_Http::make_request` for network issues. 

I think we should try/catch this exception here. Otherwise, the execution can not reach this code to allow retries. 

I tested this by: 

- Enable both client and server. 
- Client: Add products into cart, visit checkout page. 
- Server: Disable the server. 
- Client: Try to make a payment on the checkout page. 
- Client: check the WCPayment log file for today under `wp-content/uploads/wc-logs/` or ` docker/wordpress/wp-content/uploads/wc-logs/` for Docker. 
- There is only one log entry `ERROR HTTP_REQUEST_ERROR` - [code ref](https://github.com/Automattic/woocommerce-payments/blob/485926e28a9618470a9ab3f4ebc97f520c15be3b/includes/wc-payment-api/class-wc-payments-http.php#L111-L111)

Reply:
Adding a try/catch seems to work well. With steps above, there are 4 log entries `ERROR HTTP_REQUEST_ERROR`. Here is my attempt https://gist.github.com/htdat/a50eb9310ae8cd8e5a01e9be27ba3367

Reply:
Thanks for checking this, @htdat! I'll add a try/catch loop and additional tests.

However, my concern is that `Connection_Exception` can be thrown in other cases too, where response code is present. It'd be prudent to not retry if response contains an error, but it adds complexity to checks.

What's worth trying is to modify `make_request` to return different HTTP code for empty response, e.g. 503. This will allow to distinguish errors in the retry loop.

Reply:
Addressed in fb33174b

Reply:
> What's worth trying is to modify make_request to return different HTTP code for empty response, e.g. 503. This will allow to distinguish errors in the retry loop.

At first, I thought it's a good idea. However, when looking around the code you've made for `make_request` in https://github.com/Automattic/woocommerce-payments/commit/fb33174b81bcb8c3190f3b5c09d04e30bac811dc, it seems that `$exception_http_code` is always `503`. The reason is that if `is_array( $response ) === true`, the code will not execute the `if` block in this method, hence no exception will be thrown. 

I looked further to `Automattic\Jetpack\Connection\Client::remote_request` and `Automattic\Jetpack\Connection\Client::_wp_remote_request`, then finally `wp_remote_request`. If I understand correctly, it's really a network issue if these functions return a WP_Error response, i.e. a `Connection_Exception` throw in `make_request`. 

In other words, I think we can simplify the change by catching `Connection_Expection` only, which is enough to cover network issues. WDYT?

Reply:
I agree, that the check I have for `$exception_http_code` looks fragile. However, the problem is that any filter can return `WP_Error` as well, even after request was successful, reducing the reliability of this check.

I really don't like the idea of sending sensitive payment data through the code stack with so many injection points allowing any plugin to read and even change information being sent. I agree to have a check on `is_wp_error()` alone, if we agree to implement a  separate `request` method for API calls which wouldn't allow for any interventions other than specified by the caller.

Reply:
Chanaged as @htdat proposed in 44fbe8cb

Reply:
I agree that at any point, other plugins/code can add a filter to return `WP_Error` and it can be annoying for a payment request. This is a nature of WordPress with many hooks like that. Someone can argue it's good for extensibility but it also creates a lot of headache if someone abuses it. 

Even we have a separate `request` method but still use `wp_remote_request`, which has filter `http_response` to modify any response, someone can still filter and change the `response`. That said, we might need to live with it at the moment and may move this discussion to a broader audience.

> https://github.com/Automattic/woocommerce-payments/commit/44fbe8cbf930b620195eff5663d15e3d986c9fed

LGTM! Though I have some minor comments regarding tests.